reV/bespoke/place_turbines.py

```diff
@@ -0,0 +1,234 @@
+        mult = 1.0
+        while nturbs > 300:
+            packing.clear()
+            packing.min_spacing = self.min_spacing * mult
+            packing.pack_turbines_poly()
+            nturbs = len(packing.turbine_x)
+            mult *= 1.1
+        self.x_locations = packing.turbine_x
+        self.y_locations = packing.turbine_y
+
+    def optimization_objective(self, x):
+        """The optimization objective used in the bespoke optimization
+        """
+        x = [bool(y) for y in x]
+        n_turbines = np.sum(x)
+        self.wind_plant["wind_farm_xCoordinates"] = self.x_locations[x]
+        self.wind_plant["wind_farm_yCoordinates"] = self.y_locations[x]
+
+        system_capacity = n_turbines * self.turbine_capacity
+        self.wind_plant["system_capacity"] = system_capacity
+
+        self.wind_plant.assign_inputs()
+        self.wind_plant.execute()
+        aep = self.wind_plant.annual_energy()
+        cost = eval(self.cost_function, globals(), locals())
+        objective = eval(self.objective_function, globals(), locals())
+
+        return objective
+
+    def optimize(self):
+        """use a genetic algorithm to optimize wind plant layout for the user
+        defined objective function.
+        """
+        nlocs = len(self.x_locations)
+        bits = np.ones(nlocs, dtype=int)
+        bounds = np.zeros((nlocs, 2), dtype=int)
+        bounds[:, 1] = 2
+        variable_type = np.array([])
+        for _ in range(nlocs):
+            variable_type = np.append(variable_type, "int")
+        ga = GeneticAlgorithm(bits, bounds, variable_type,
+                              self.optimization_objective)
+
+        ga.max_generation = 10000
+        ga.population_size = 25
+        ga.crossover_rate = 0.2
+        ga.mutation_rate = 0.01
+        ga.tol = 1E-6
+        ga.convergence_iters = 10000
+        ga.max_time = self.ga_time
+        ga.optimize_ga()
+
+        optimized_design_variables = ga.optimized_design_variables
+        optimized_design_variables = \
+            [bool(y) for y in optimized_design_variables]
+
+        self.objective = 0.0
+
+        self.turbine_x = self.x_locations[optimized_design_variables]
+        self.turbine_y = self.y_locations[optimized_design_variables]
+        self.nturbs = np.sum(optimized_design_variables)
+        self.capacity = self.turbine_capacity * self.nturbs
+        self.area = self.full_polygons.area
+        if self.area != 0.0:
+            self.capacity_density = self.capacity / self.area * 1E3
+        else:
+            self.capacity_density = 0.0
+
+        self.wind_plant["wind_farm_xCoordinates"] = \
+            self.x_locations[optimized_design_variables]
+        self.wind_plant["wind_farm_yCoordinates"] = \
+            self.y_locations[optimized_design_variables]
+        self.wind_plant["system_capacity"] = self.capacity
+        self.wind_plant.assign_inputs()
+        self.wind_plant.execute()
+        self.aep = self.wind_plant.annual_energy()
```

Comment:
I would recommend that all variables from `self.turbine_x` on line 205 to `self.aep` on line 222 should actually be `@property` objects that retrieve their values on-the-fly instead of being assigned in this function call. Dynamically retrieved properties will ensure that these values are always current and are always self-consistent. It also allows you to add a small docstring describing each attribute which is really nice for documentation.

Reply:
We'll need to talk about this, all/most of the variables are meaningless before running the optimization and are assigned with the optimal values of the optimization.

Reply:
All of these variables are now accessed through attributes, and return None if they have not yet been assigned